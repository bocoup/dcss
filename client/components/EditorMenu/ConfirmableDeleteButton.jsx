import React from 'react';
import PropTypes from 'prop-types';
import { Confirm, Icon, Menu, Popup } from 'semantic-ui-react';

class ConfirmableDeleteButton extends React.Component {
    constructor(props) {
        super(props);
        this.state = {
            open: false
        };

        this.onClick = this.onClick.bind(this);
        this.onCancel = this.onCancel.bind(this);
        this.onComplete = this.onCancel;
    }

    onClick() {
        this.setState({ open: true });
    }

    onCancel() {
        this.setState({ open: false });
    }

    render() {
        const {
            onCancel,
            onClick,
            onComplete,
            props: { itemType, onConfirm }
        } = this;

        const ariaLabel = `Delete ${itemType}`.trim();
        const content = itemType
            ? `Are you sure you want to delete this ${itemType}?`
            : `Are you sure you want to delete?`;

        return (
            <React.Fragment>
                <Popup
                    content={ariaLabel}
                    trigger={
                        <Menu.Item onClick={onClick}>
                            <Icon name="trash" aria-label={ariaLabel} />
                        </Menu.Item>
                    }
                />
                <Confirm
                    open={this.state.open}
                    content={content}
                    cancelButton="No Thanks"
                    confirmButton="Confirm Delete"
                    onCancel={onCancel}
                    onConfirm={() => {
                        onConfirm();
                        onComplete();
                    }}
                />
            </React.Fragment>
        );
    }
}

ConfirmableDeleteButton.propTypes = {
    itemType: PropTypes.string,
    onConfirm: PropTypes.func.isRequired
};
export default ConfirmableDeleteButton;
