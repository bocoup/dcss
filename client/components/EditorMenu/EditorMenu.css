.em__height {
  min-height: 45px !important;
}

.em__icon-group-margin {
  margin-right: 0.5rem!important;
}

.em__search-input-box-right {
  padding-right: 0 !important;
}

.em__sticky {
  z-index: 1;
  background: white;
  position: -webkit-sticky;
  position: sticky;
  top: 0;
}


.em__fullwidth /* minus left and right padding + margin */ {
  width: calc(100vw - 30px) !important;
}

.ui.segment > .em__fullwidth /* minus left and right padding + margin */ {
  width: calc(100vw - 30px) !important;
}


@media only screen and (max-width: 767px) {
  .em__height input {
    width: 12rem;
  }
}


.em__overflow-truncated {
  overflow: hidden !important;
  white-space: nowrap !important;
  text-overflow: ellipsis !important;
}
