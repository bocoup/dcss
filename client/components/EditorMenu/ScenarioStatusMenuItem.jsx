import React from 'react';
import PropTypes from 'prop-types';
import changeCase from 'change-case';
import { Dropdown, Menu, Icon } from '@components/UI';

let cachedStatusOptions = [];

class ScenarioStatusMenuItem extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
      statusOptions: cachedStatusOptions
    };

    if (!cachedStatusOptions) {
      this.fetchStatusOptions();
    }
  }

  async fetchStatusOptions() {
    const statusOptions = await (cachedStatusOptions ||
      (await fetch('/api/status')).json());

    statusOptions.forEach(
      statusOption =>
        (statusOption.name = changeCase.titleCase(statusOption.name))
    );

    cachedStatusOptions = statusOptions; // eslint-disable-line require-atomic-updates

    this.setState({ statusOptions });
  }

  getStatusDisplay({ asIcon = false }) {
    const status = this.state.statusOptions.find(
      status => status.id === this.props.status
    );

    const color = {
      1: 'pink',
      2: 'green',
      3: 'orange'
    }[status.id];

    return asIcon ? (
      <Icon corner="top right" name="check" color={color} />
    ) : (
      `${status.name} (${status.description})`
    );
  }

  render() {
    if (!this.state.statusOptions.length) {
      return null;
    }

    const statusAsIcon = this.getStatusDisplay({ asIcon: true });
    const statusAsText = this.getStatusDisplay({ asIcon: false });
    const dropdownTrigger = (
      <React.Fragment>
        <Icon.Group className="em__icon-group-margin">
          <Icon name="pencil alternate" />
          {statusAsIcon}
        </Icon.Group>
        {statusAsText}
      </React.Fragment>
    );
    return (
      <Menu.Menu key="menu-item-scenario-status">
        <Dropdown item text={dropdownTrigger}>
          <Dropdown.Menu>
            {this.state.statusOptions.map(({ id, name, description }) => (
              <Dropdown.Item
                name="save-status"
                id={id}
                key={id}
                onClick={this.props.onClick}
              >
                {name} ({description})
              </Dropdown.Item>
            ))}
          </Dropdown.Menu>
        </Dropdown>
      </Menu.Menu>
    );
  }
}

ScenarioStatusMenuItem.propTypes = {
  // scenario: PropTypes.object,
  status: PropTypes.number,
  onClick: PropTypes.func.isRequired
};

export default ScenarioStatusMenuItem;
