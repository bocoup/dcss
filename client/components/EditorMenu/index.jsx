import React from 'react';
import PropTypes from 'prop-types';
import { Icon, Menu, Popup } from '@components/UI';
import ConfirmableDeleteButton from './ConfirmableDeleteButton';
import './EditorMenu.css';

export default class EditorMenu extends React.Component {
  shouldComponentUpdate(newProps) {
    if (newProps.isDragging) {
      return false;
    }
    return true;
  }

  render() {
    const {
      className,
      draghandle = {},
      type,
      items,
      isDragging,
      ...props
    } = this.props;
    let menuClassName = 'em__height';
    if (className) {
      menuClassName += ` ${className}`;
    }

    void isDragging;
    // const menuItems = Object.entries(items).reduce((accum, [prop, item]) => {
    //   if (Array.isArray(item)) {
    //     accum.push(...item.filter(x => x));
    //   }
    // }, []);
    // console.log(menuItems);

    return (
      <Menu
        {...props}
        {...draghandle}
        icon
        borderless
        className={menuClassName}
      >
        {items.left && (
          <React.Fragment>
            {items.left
              .filter(item => item)
              .map((item, index) => {
                return item.props.name ? (
                  <Popup
                    size="small"
                    key={index}
                    content={item.props.name}
                    trigger={item}
                  />
                ) : (
                  item
                );
              })}
          </React.Fragment>
        )}
        {items.save && (
          <Popup
            size="small"
            content={`Save ${type}`}
            trigger={
              <Menu.Item
                aria-label={`Save ${type}`}
                disabled={items.save.disabled}
                name={`save-${type}`}
                onClick={items.save.onClick}
              >
                <Icon name="save outline" />
              </Menu.Item>
            }
          />
        )}
        {items.delete && (
          <ConfirmableDeleteButton
            aria-label={`Delete ${type}`}
            disabled={items.delete.disabled}
            name={`delete-${type}`}
            itemType={type}
            onConfirm={items.delete.onConfirm}
          />
        )}

        {items.right && (
          <React.Fragment>
            {items.right
              .filter(item => item)
              .map((item, index) => {
                return item.props.name ? (
                  <Popup
                    size="small"
                    key={index}
                    content={item.props.name}
                    trigger={item}
                  />
                ) : (
                  item
                );
              })}
          </React.Fragment>
        )}
      </Menu>
    );
  }
}

const VALID_PROPS = ['delete', 'editable', 'left', 'right', 'save'];
EditorMenu.propTypes = {
  pointing: PropTypes.bool,
  secondary: PropTypes.bool,
  className: PropTypes.string,
  draghandle: PropTypes.object,
  isDragging: PropTypes.bool,
  index: PropTypes.number,
  items: function(props, propName) {
    const { items } = props;
    if (propName === 'items') {
      if (!Object.keys(items).every(v => VALID_PROPS.includes(v))) {
        return new Error('EditorMenu: Invalid item property');
      }

      if (items.save) {
        if (
          !Object.keys(items.save).every(v =>
            ['disabled', 'onClick'].includes(v)
          )
        ) {
          return new Error('EditorMenu: Invalid items.save property');
        }
      }

      if (items.delete) {
        if (
          !Object.keys(items.delete).every(v =>
            ['disabled', 'onConfirm'].includes(v)
          )
        ) {
          return new Error('EditorMenu: Invalid item.delete property');
        }
      }

      if (items.left && !Array.isArray(items.left)) {
        return new Error('EditorMenu: Invalid items.left property');
      }

      if (items.right && !Array.isArray(items.right)) {
        return new Error('EditorMenu: Invalid items.right property');
      }
    }

    return null;
  },
  type: PropTypes.string
};
