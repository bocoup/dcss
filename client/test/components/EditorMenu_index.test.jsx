import {
  fetchImplementation,
  mounter,
  reduxer,
  snapshot,
  state,
} from '../bootstrap';
import {
  unmountComponentAtNode
} from 'react-dom';
import {
  cleanup,
  fireEvent,
  render
} from '@testing-library/react';
import EditorMenu from '../../components/EditorMenu/index.jsx';

beforeAll(() => {
  (window || global).fetch = jest.fn();
});

let container = null;
beforeEach(() => {
  container = document.createElement('div');
  container.setAttribute('id', 'root');
  document.body.appendChild(container);

  fetchImplementation(fetch);
});

afterEach(() => {
  unmountComponentAtNode(container);
  container.remove();
  container = null;
});

const sharedProps = {
  history: {
    push() {},
  },
};

test('EditorMenu', () => {
  expect(EditorMenu).toBeDefined();
});

test('Snapshot 1', () => {
  const props = {
    ...sharedProps,
    items: {
      delete: {
        onConfirm() {}
      },
      editable: null,
      left: null,
      right: null,
      save: {
        onClick() {}
      },
    },
  };
  const mounted = mounter(reduxer(EditorMenu, props, state))();
  expect(snapshot(mounted)).toMatchSnapshot();
});

/*{INJECTION}*/
