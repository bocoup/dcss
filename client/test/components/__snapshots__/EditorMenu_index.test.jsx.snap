// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot 1 1`] = `"<div history=\\"[object Object]\\" class=\\"ui borderless icon em__height menu\\"><a tabindex=\\"0\\" aria-label=\\"Save undefined\\" class=\\"item\\"><i aria-hidden=\\"true\\" class=\\"save outline icon\\"></i></a><a tabindex=\\"0\\" class=\\"item\\"><i aria-label=\\"Delete this undefined\\" class=\\"trash alternate outline icon\\"></i></a></div>"`;
